### PR TITLE
Port IPP SHA384 and SM3 algorithms to SlimBootloader

### PR DESCRIPTION
**Pull request by lsubash**
SHA512_384 and SM3 crypto functions were referenced from
https://github.com/intel/ipp-crypto.git

Selectively taken Ipp Hash methods

Signed-off-by: Subash Lakkimsetti <subashx.lakkimsetti@intel.com>

### CODE REVIEW

**Comment by lsubash** [comment]
No Size impact for APL with/without this patch.

Stage1A: EFI_FV_TAKEN_SIZE = 0x4060
Srage1B: EFI_FV_TAKEN_SIZE = 0xf170
Stage2: EFI_FV_TAKEN_SIZE = 0x29ec0

**Comment by mauricema** [comment]
There are build failure,  are you going to fix it ?

**Comment by sagraw2** [comment]
This looks ok to me. We should probably merge this and then push the optimization patch worked by Maurice on top of it.

**Comment by myguitar** [comment]
Can you run 'python BaseTools/Scripts/PatchCheck.py'?

File: BootloaderCommonPkg/Library/IppCryptoLib/auth/ippcp.h [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /** @file
 
-  Copyright (c) 2018, Intel Corporation. All rights reserved.<BR>
+  Copyright (c) 2019, Intel Corporation. All rights reserved.<BR>
```
Comment (myguitar) [inline BootloaderCommonPkg/Library/IppCryptoLib/auth/ippcp.h]: 2018 - 2019

File: BootloaderCommonPkg/Library/IppCryptoLib/auth/owncp.h [hunk 2]
```diff
@@ -1,6 +1,6 @@
 /** @file
 
-  Copyright (c) 2018, Intel Corporation. All rights reserved.<BR>
+  Copyright (c) 2019, Intel Corporation. All rights reserved.<BR>
```
Comment (myguitar) [inline BootloaderCommonPkg/Library/IppCryptoLib/auth/owncp.h]: 2018 - 2019

File: BootloaderCommonPkg/Library/IppCryptoLib/auth/pcpsha512ca.c [hunk 3]
```diff
@@ -0,0 +1,567 @@
+/** @file
+
+  Copyright (c) 2018-2019, Intel Corporation. All rights reserved.<BR>
```
Comment (myguitar) [inline BootloaderCommonPkg/Library/IppCryptoLib/auth/pcpsha512ca.c]: This is a new file, right? 2018-2019 -> 2019

File: BootloaderCommonPkg/Library/IppCryptoLib/auth/pcpsm3ca.c [hunk 4]
```diff
@@ -0,0 +1,585 @@
+/** @file
+
+  Copyright (c) 2018-2019, Intel Corporation. All rights reserved.<BR>
```
Comment (myguitar) [inline BootloaderCommonPkg/Library/IppCryptoLib/auth/pcpsm3ca.c]: 2018-2019 -> 2019

**Pull request merged**